upgrade_guide/migrating_from_astyanax/language_level_changes/README.md

```diff
@@ -0,0 +1,66 @@
+             |                                                                                            |
+One Thrift   |                                                                                           -/
+COLUMNFAMILY |          
+             |  
+             |  /------------\              /---------------+---------------+---------+
+             |  |            |              |     Col 1     |     Col 4     |         |
+             |  | Key : 2    | ---------->  +---------------+---------------+ ...     |
+             |  |            |              |     Val : 'a' |     Val : 'b' |         |
+             |  \------------/              \---------------+---------------+---------+
+             |  
+             \- 
+
+
+                      -----------------------------------------------------------------------
+
+
+                                                     CQL
+                                                     
+             /-                              
+             |                               
+             |  /--------------------+---------------------------+-----------------------+----------------------+-----------------------------\
+             |  |       key          |           Col1            |         Col2          |         Col3         |            Col4             |  
+             |  +--------------------+---------------------------+-----------------------+----------------------+-----------------------------+ -\
+             |  | cRED   1           |   c1AB     'a'            | cFA0     'b'          |          'c'         |             null            |  +--> One CQL
+   One CQL   |  +--------------------+---------------------------+-----------------------+----------------------+-----------------------------+ -/    ROW
+   TABLE     |  |        2           |             'a'           |          null         |          null        |             'b'             |
+             |  +--------------------+---------------------------+-----------------------+----------------------+-----------------------------+
+             |  |       ...          |            ...            |          ...          |          ...         |             ...             |
+             |  +--------------------+---------------------------+-----------------------+----------------------+-----------------------------+
+             \-
+```
+
+Some of the columns of a *CQL Table* have a special role that is specifically
+related to the *Cassandra* architecture. Indeed, the *Row key* of the *Thrift Row*,
+becomes the *Partition Key* in the *CQL Table*, and can be composed of 1 or multiple
+*CQL columns* (the key column in Figure 1). The *“Column”* part of the Column-value
+component in a *Thrift Row*, becomes the *Clustering ColumnKey* in *CQL*, and can
+also be composed of multiple columns (in the figure, column1 is the only column 
+composing the *Clustering ColumnKey*).
```

Comment:
Sorry, it's easy to conflate terms so i'm probably confusing things a bit.

Taking a step back to describe what I'm trying to say with an example data model.   I was thinking of the table that you were sharing having a composite comparator that was `CompositeType(Int32, UTF8Type)`  where `Int32` is the value of `Col1`, and `UTF8Type` is the name of the column.

As an example, say you want to represent readings from a thermostat.  The thermostat is identified by an integer id which will be our row key. A thermostat has 'readings' identified by a reading id, and each reading has a `temp` (for the current temperature), a `mode` (cooling, heating, off) and a setpoint (for target temperature).

Using cassandra-cli, i could create a schema like so:

```
create keyspace ks;
use ks;
create column family readings
with key_validation_class = Int32Type
and comparator = 'CompositeType(Int32Type, UTF8Type)'
and default_validation_class = UTF8Type;
```

I could get even fancier and provide column metadata, but leaving this as is to keep it simple.

I could then create multiple readings for an individual thermostat, i.e.:

```
set readings[0]['0:temp'] = '72';
set readings[0]['0:mode'] = 'heating';
set readings[0]['0:setpoint'] = '73';
set readings[0]['1:temp'] = '75';
set readings[0]['1:mode'] = 'cooling';
set readings[0]['1:setpoint'] = '70';
set readings[1]['5:temp'] = '65';
set readings[1]['5:mode'] = 'off';
set readings[1]['5:setpoint'] = '70';
```

When reading back the data, I get:

```
[default@ks] get readings[0];
=> (name=0:mode, value=heating, timestamp=1474483110889000)
=> (name=0:setpoint, value=73, timestamp=1474483119088000)
=> (name=0:temp, value=72, timestamp=1474483101433000)
=> (name=1:mode, value=cooling, timestamp=1474483137460000)
=> (name=1:setpoint, value=70, timestamp=1474483147380000)
=> (name=1:temp, value=75, timestamp=1474483127437000)
[default@ks] get readings[1];
=> (name=5:mode, value=off, timestamp=1474483198322000)
=> (name=5:setpoint, value=70, timestamp=1474483205968000)
=> (name=5:temp, value=65, timestamp=1474483192274000)
```

In that example, for a Thrift row for a given key,  I have multiple readings.  In CQL, The concept of a row layout would depend on the clustering, but I would preferably model this like:

```
create table readings_cql (k int, reading_id int, mode text, setpoint float, temp float, primary key (k, reading_id));

-- insert data
insert into readings_cql (k, reading_id, mode, setpoint, temp) values (0, 0, 'heating', 73, 72);
insert into readings_cql (k, reading_id, mode, setpoint, temp) values (0, 1, 'cooling', 70, 75);
insert into readings_cql (k, reading_id, mode, setpoint, temp) values (1, 5, 'off', 70, 65);
```

This would give me back:

```
select * from readings_cql;

 k | reading_id | mode    | setpoint | temp
---+------------+---------+----------+------
 1 |          5 |     off |       70 |   65
 0 |          0 | heating |       73 |   72
 0 |          1 | cooling |       70 |   75
```

This looks more like your second (current) diagram.

Where as the thrift table represented as a CQL table looks more like your first (https://gist.github.com/newkek/4a0cbe91577886383aaa9ef89701cf03)

```
cqlsh> select * from ks.readings;

 key | column1 | column2  | value
-----+---------+----------+---------
   1 |       5 |     mode |     off
   1 |       5 | setpoint |      70
   1 |       5 |     temp |      65
   0 |       0 |     mode | heating
   0 |       0 | setpoint |      73
   0 |       0 |     temp |      72
   0 |       1 |     mode | cooling
   0 |       1 | setpoint |      70
   0 |       1 |     temp |      75
```

So what the `One CQL Table` will look like will ultimately depend on the CQL schema.  Ultimately the stored data is similar, it's just how CQL determines what is a 'Row' depends on the clustering.

Comparing the data for readings (thrift) and readings_cql, the layout on disk is virtually the same (other than the values because my value type is different, and the ghost cell for CQL):

```
Thrift table:
[default@ks] get readings[0];
=> (name=0:mode, value=heating, timestamp=1474483110889000)
=> (name=0:setpoint, value=73, timestamp=1474483119088000)
=> (name=0:temp, value=72, timestamp=1474483101433000)
=> (name=1:mode, value=cooling, timestamp=1474483137460000)
=> (name=1:setpoint, value=70, timestamp=1474483147380000)
=> (name=1:temp, value=75, timestamp=1474483127437000)

CQL table:
[default@ks] get readings_cql[0];
=> (name=0:, value=, timestamp=1474483980386000)
=> (name=0:mode, value=68656174696e67, timestamp=1474483980386000)
=> (name=0:setpoint, value=42920000, timestamp=1474483980386000)
=> (name=0:temp, value=42900000, timestamp=1474483980386000)
=> (name=1:, value=, timestamp=1474483996207000)
=> (name=1:mode, value=636f6f6c696e67, timestamp=1474483996207000)
=> (name=1:setpoint, value=428c0000, timestamp=1474483996207000)
=> (name=1:temp, value=42960000, timestamp=1474483996207000)
```

With all that defined, what was confusing to me is the Thrift Diagram (`One Thrift COLUMNFAMILY`) which is the same for both diagrams.

In your diagram you have three colors:

1. red: Thrift: Row Key, CQL: Partition Key
2. blue: Thrift: Column Comparator, CQL: Clustering Column Value
3. orange: Thrift: Column Value, CQL: Non-Clustering Column Value

If I'm interpreting that correctly and 'Col1' is a clustering column, it wouldn't be represented as its own column for Thrift.  Instead if would be part of the column name in Thrift diagram where it would not for the CQL diagram.  

I know that is maybe i'm getting too semantical and thinking too much about the storage model, but I think it is hard to divorce thrift from the storage model, as to me they are the same.  [One of the older thrift -> CQL guides](http://www.datastax.com/dev/blog/thrift-to-cql3) uses the storage model to describe CQL and how different column family schemas would be modeled with CQL:

> CQL3 (the Cassandra Query Language) provides a new API to work with Cassandra. Where the legacy thrift API exposes the internal storage structure of Cassandra pretty much directly, CQL3 provides a thin abstraction layer over this internal structure.

Hopefully that makes sense, I had to revisit thrift to level set my thoughts.